Add an integration guide Test Filter Standardization Readme (#5186)
ROCm/rocm-libraries: shared/ctest/README.md

```cmake
# Filter format: positive_string-exclude_string:gpu_exclude_string
add_test(
  NAME miopen_gtest_standard_gfx1150_suite
  COMMAND miopen_gtest --gtest_filter="*Fusion*:*Conv*-*DeepBench*:*Slow*:*gfx942*"
)
```

## Integration Guide

> This guide walks through integrating the shared YAML-based test categorization into a GTest based ROCm library project, using MIOpen as the reference implementation.

Reviewer: ```suggestion
This guide walks through integrating the shared YAML-based test categorization into a GTest-based ROCm library project, using MIOpen as the reference implementation.
```

## Prerequisites

- A monolithic GTest binary (single executable containing all tests)

> - Python 3 available in the build environment

Reviewer: ```suggestion
- Python 3 which is available in the build environment
```

  - "*UnsupportedOnGfx11*"

exclude_gpu:
  exclude_gpu_gfx11X:
    # Applies to all gfx11xx GPUs (gfx1100, gfx1150, gfx1151, etc.)
    test_patterns: *gpu_common_patterns
    labels:
      - "quick"        # Must list every category this exclusion applies to
      - "standard"
      - "comprehensive"
      - "full"
      - "ex_gpu_gfx11X"  # Required: ex_gpu_<arch> label for CTest filtering

  exclude_gpu_gfx1150:
    # Additional exclusions specific to gfx1150 (on top of gfx11X)
    test_patterns:
      - "*Gfx1150SpecificFailure*"
    labels:
      - "quick"
      - "ex_gpu_gfx1150"

  exclude_gpu_gfx950:
    test_patterns:
      - "*Gfx950Issue*"
    labels:
      - "quick"
      - "standard"
      - "ex_gpu_gfx950"
```

**How hierarchical matching works for the above snippet:**

| GPU on machine | Exclusion configs applied                    |
| -------------- | -------------------------------------------- |
| gfx1100        | `exclude_gpu_gfx11X`                         |
| gfx1150        | `exclude_gpu_gfx11X` + `exclude_gpu_gfx1150` |
| gfx1151        | `exclude_gpu_gfx11X`                         |
| gfx950         | `exclude_gpu_gfx950`                         |

> **Important:** The `labels` list in each `exclude_gpu` entry must include every test category that the exclusion applies to. The parser matches these labels against `category_data` keys to determine which categories get the GPU-specific test variants.

Reviewer: ```suggestion
**Important:** The `labels` list in each `exclude_gpu` entry must include every test category that the exclusion applies to. The parser matches these labels against `category_data` keys to determine which categories receive the GPU-specific test variants.
```

## Step 2: Integrate in CMakeLists.txt

### Enable CTest

Ensure `enable_testing()` is called somewhere in your CMake hierarchy before `apply_test_category_labels()`. Without it, `add_test()` calls have no effect and CTest won't discover any tests. This is typically done in the project's test-level `CMakeLists.txt`:

```cmake
if(BUILD_TESTING)
    enable_testing()
endif()
```

If your project already calls `enable_testing()` (or uses `include(CTest)`, which calls it internally), no additional change is needed.

### Integration with Install Support

> For projects that distribute test binaries (e.g., via TheRock), generate an install-time CTestTestfile.cmake, so that it can be installed to wherever the final test binaries are present.

Reviewer: ```suggestion
For projects that distribute test binaries (for example, via TheRock), generate an install-time CTestTestfile.cmake, so that it can be installed wherever the final test binaries are present.
```

```cmake
set(ROCM_LIBRARIES_ROOT ${CMAKE_CURRENT_SOURCE_DIR}/../../..)
include(${ROCM_LIBRARIES_ROOT}/shared/ctest/TestCategories.cmake)

# Create install-time test file
set(INSTALL_TEST_FILE "${CMAKE_CURRENT_BINARY_DIR}/install_CTestTestfile.cmake")
file(WRITE "${INSTALL_TEST_FILE}"
[=[
# Generated test definitions for install-time execution.
# Tests use relative paths to the test binary.
]=]
)

# Apply categorization (4th argument enables install file generation)
if(EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/test_categories.yaml")
    apply_test_category_labels(
        my_project_gtest
        "${CMAKE_CURRENT_SOURCE_DIR}/test_categories.yaml"
        "${PROJECT_BINARY_DIR}"
        "${INSTALL_TEST_FILE}"
    )
endif()

# Install the generated CTestTestfile.cmake alongside the test binary
install(
    FILES "${INSTALL_TEST_FILE}"
    DESTINATION "${CMAKE_INSTALL_BINDIR}/${PROJECT_NAME}"
    COMPONENT tests
    RENAME "CTestTestfile.cmake"
)
```

### Excluding Category Suites from `make check`

> All YAML-generated tests are named with a `_suite` suffix. If your project has a `check` target that runs `ctest`, the newly generated tests may be excluded to avoid extended run times:

Reviewer: ```suggestion
All YAML-generated tests are named with a `_suite` suffix. If your project has a `check` target that runs `ctest`, the newly generated tests can be excluded to avoid extended run times:
```

cmake -DBUILD_TESTING=ON ..
cmake --build .

# List all registered tests
ctest -N

# Run a specific category (no GPU exclusions)
ctest -L quick -LE ex_gpu

# Run a specific category with GPU-specific exclusions
ctest -L quick -L ex_gpu_gfx1150

# Run with verbose output
ctest -L standard -L ex_gpu_gfx950 -V
```

## Generated Test Names

The framework generates tests with predictable naming:

| Type           | Name Format                            | Example                                |
| -------------- | -------------------------------------- | -------------------------------------- |
| Category suite | `{target}_{category}_suite`            | `my_project_gtest_quick_suite`         |
| GPU exclusion  | `{target}_{category}_{gpu_arch}_suite` | `my_project_gtest_quick_gfx1150_suite` |

## Generated Labels

Each test gets labels that enable flexible CTest filtering:

| Test Type      | Labels                                                              |
| -------------- | ------------------------------------------------------------------- |
| Category suite | Category labels from YAML (e.g., `quick`, `standard`)               |

> | GPU exclusion  | Category labels + `ex_gpu_<arch>` (e.g., `quick`, `ex_gpu_gfx1150`) |

Reviewer: ```suggestion
| Category suite | Category labels from YAML (for example, `quick`, `standard`)               |
| GPU exclusion  | Category labels + `ex_gpu_<arch>` (for example, `quick`, `ex_gpu_gfx1150`) |
```

## Function Reference

### `apply_test_category_labels(target_name yaml_file working_dir [install_test_file])`

| Parameter           | Required | Description                                                                                         |
| ------------------- | -------- | --------------------------------------------------------------------------------------------------- |
| `target_name`       | Yes      | Name of the GTest executable (must not be empty)                                                   |
| `yaml_file`         | Yes      | Absolute path to `test_categories.yaml` (must exist)                                               |
| `working_dir`       | Yes      | Working directory for test execution (must be a valid directory)                                   |
| `install_test_file` | No       | Path to the generated install-time `CTestTestfile.cmake`. Not required for local builds, but      |

> |                     |          | required for install/package workflows (e.g., TheRock).                                            |

Reviewer: ```suggestion
|                     |          | required for install/package workflows (for example, TheRock).                                            |
```

```

## Checklist

- Created `test_categories.yaml` in your project's test directory
- At least one category has `test_patterns` and `labels` defined
- `ROCM_LIBRARIES_ROOT` is set correctly (points to the monorepo root)
- `include(${ROCM_LIBRARIES_ROOT}/shared/ctest/TestCategories.cmake)` added before calling the function
- `apply_test_category_labels()` called with the correct test executable name
- GPU exclusion `labels` lists include every applicable category name
- Each `exclude_gpu_<arch>` entry has an `ex_gpu_<arch>` label
- `ctest -N` shows the expected test suites after building

## Reference Implementation

See the MIOpen integration for a complete working example:

- **YAML config:** [projects/miopen/test/gtest/test_categories.yaml](../../projects/miopen/test/gtest/test_categories.yaml)
- **CMake integration:** [projects/miopen/test/gtest/CMakeLists.txt](../../projects/miopen/test/gtest/CMakeLists.txt)
- **Shared module:** [shared/ctest/TestCategories.cmake](./TestCategories.cmake)
- **Parser script:** [shared/ctest/parse_test_categories.py](./parse_test_categories.py)
=======
### **Install-time CTestTestfile (TheRock / install tree)**

> When tests are installed (e.g. into `/opt/rocm/bin/`), the **build-tree** test definitions are not installed. To run CTest from the **installed** location (e.g. on TheRock or any system that only has the install tree), projects can generate an **install-time CTestTestfile** that uses **relative paths** to the test executable.

Reviewer: replace **e.g.** with **for example** everywhere. 
it should be followed with a comma

Author: Done!
